File: backend/app/schemas/automation.py
import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.constants import (
    AutomationActionType,
    AutomationApprovalStatus,
    AutomationExecutionStatus,
    AutomationSuggestionStatus,
    AutomationTargetType,
)


class AutomationPreview(BaseModel):
    title: str
    body: str
    content_hash: str
    external_mutation: Literal[False] = False


class AutomationSuggestionResponse(BaseModel):
    id: uuid.UUID
    workspace_id: uuid.UUID
    target_type: AutomationTargetType | str
    target_id: uuid.UUID
    role_id: uuid.UUID | None = None
    application_id: uuid.UUID | None = None
    artifact_id: uuid.UUID | None = None
    action_type: AutomationActionType | str
    title: str
    summary: str
    reason: str
    basis: str
    confidence: str
    source_inputs: dict[str, Any] = Field(default_factory=dict)
    preview: AutomationPreview
    preview_hash: str
    status: AutomationSuggestionStatus | str
    expires_at: datetime | None = None
    policy_version: str
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class AutomationSuggestionListResponse(BaseModel):
    workspace_id: uuid.UUID | None = None
    target_type: AutomationTargetType | str | None = None
    target_id: uuid.UUID | None = None
    suggestions: list[AutomationSuggestionResponse] = Field(default_factory=list)
    external_actions_enabled: Literal[False] = False


class AutomationDecisionRequest(BaseModel):
    reason: str | None = Field(default=None, max_length=2000)
    actor: Literal["user"] = "user"


class AutomationApprovalLogResponse(BaseModel):
    id: uuid.UUID
    workspace_id: uuid.UUID
    suggestion_id: uuid.UUID
    actor: str
    target_type: AutomationTargetType | str
    target_id: uuid.UUID
    action_type: AutomationActionType | str
    preview: AutomationPreview
    preview_hash: str
    approval_status: AutomationApprovalStatus | str
    dismissal_or_rejection_reason: str | None = None
    execution_status: AutomationExecutionStatus | str
    execution_result: dict[str, Any] = Field(default_factory=dict)
    external_mutation: Literal[False] = False
    policy_version: str
    created_at: datetime
    decided_at: datetime | None = None
    executed_at: datetime | None = None


class AutomationApprovalLogListResponse(BaseModel):
    workspace_id: uuid.UUID | None = None
    logs: list[AutomationApprovalLogResponse] = Field(default_factory=list)


class AutomationPreferencesResponse(BaseModel):
    id: uuid.UUID
    workspace_id: uuid.UUID
    enabled: bool = True
    suggestion_categories: list[AutomationActionType | str] = Field(default_factory=list)
    follow_up_suggestion_days: int = 7
    artifact_readiness_checks_enabled: bool = True
    communication_drafts_enabled: bool = True
    internal_state_change_suggestions_enabled: bool = True
    future_external_actions_enabled: Literal[False] = False
    quiet_mode: bool = False
    policy_version: str
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class AutomationPreferencesUpdate(BaseModel):
    enabled: bool | None = None
    suggestion_categories: list[AutomationActionType] | None = None
    follow_up_suggestion_days: int | None = Field(default=None, ge=1, le=90)
    artifact_readiness_checks_enabled: bool | None = None
    communication_drafts_enabled: bool | None = None
    internal_state_change_suggestions_enabled: bool | None = None
    future_external_actions_enabled: Literal[False] | None = None
    quiet_mode: bool | None = None

File: frontend/src/types/automation.ts
export type AutomationActionType =
  | "follow_up_suggestion"
  | "reminder_suggestion"
  | "artifact_readiness_check"
  | "communication_draft"
  | "workflow_state_suggestion"
  | "opportunity_review_suggestion"
  | "future_external_action_disabled";

export type AutomationTargetType =
  | "workspace"
  | "opportunity"
  | "application"
  | "artifact"
  | "reminder";

export type AutomationSuggestionStatus =
  | "active"
  | "dismissed"
  | "rejected"
  | "approved"
  | "expired";

export type AutomationApprovalStatus =
  | "pending"
  | "approved"
  | "rejected"
  | "dismissed"
  | "expired";

export type AutomationExecutionStatus =
  | "not_applicable"
  | "pending"
  | "succeeded"
  | "failed"
  | "canceled";

export interface AutomationPreview {
  title: string;
  body: string;
  content_hash: string;
  external_mutation: false;
}

export interface AutomationSuggestion {
  id: string;
  workspace_id: string;
  target_type: AutomationTargetType | string;
  target_id: string;
  role_id: string | null;
  application_id: string | null;
  artifact_id: string | null;
  action_type: AutomationActionType | string;
  title: string;
  summary: string;
  reason: string;
  basis: string;
  confidence: string;
  source_inputs: Record<string, unknown>;
  preview: AutomationPreview;
  preview_hash: string;
  status: AutomationSuggestionStatus | string;
  expires_at: string | null;
  policy_version: string;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface AutomationSuggestionListResponse {
  workspace_id: string | null;
  target_type: AutomationTargetType | string | null;
  target_id: string | null;
  suggestions: AutomationSuggestion[];
  external_actions_enabled: false;
}

export interface AutomationApprovalLog {
  id: string;
  workspace_id: string;
  suggestion_id: string;
  actor: string;
  target_type: AutomationTargetType | string;
  target_id: string;
  action_type: AutomationActionType | string;
  preview: AutomationPreview;
  preview_hash: string;
  approval_status: AutomationApprovalStatus | string;
  dismissal_or_rejection_reason: string | null;
  execution_status: AutomationExecutionStatus | string;
  execution_result: Record<string, unknown>;
  external_mutation: false;
  policy_version: string;
  created_at: string;
  decided_at: string | null;
  executed_at: string | null;
}

export interface AutomationApprovalLogListResponse {
  workspace_id: string | null;
  logs: AutomationApprovalLog[];
}

export interface AutomationPreferences {
  id: string;
  workspace_id: string;
  enabled: boolean;
  suggestion_categories: AutomationActionType[];
  follow_up_suggestion_days: number;
  artifact_readiness_checks_enabled: boolean;
  communication_drafts_enabled: boolean;
  internal_state_change_suggestions_enabled: boolean;
  future_external_actions_enabled: false;
  quiet_mode: boolean;
  policy_version: string;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface AutomationPreferencesPayload {
  enabled?: boolean;
  suggestion_categories?: AutomationActionType[];
  follow_up_suggestion_days?: number;
  artifact_readiness_checks_enabled?: boolean;
  communication_drafts_enabled?: boolean;
  internal_state_change_suggestions_enabled?: boolean;
  future_external_actions_enabled?: false;
  quiet_mode?: boolean;
}
